### PR TITLE
[Impeller] Add a Stage 2 pass to the Vulkan shader compilation pipeline.

### PR DESCRIPTION
**Pull request by chinmaygarde**
We just used to use the output of Stage 1 for Vulkan since it was already SPIRV. But that contained debugging information that bloated the size of the SPIRV and the resulting binaries. This reworks the compiler to add a second stage that strips the debug information.

While stripping the debug information is effectively what happens, I couldn't find an existing utility in the compiler toolbox that does this. So the source files are processed again for stage 2, but this time without generating the debug information.

Just a rudimentary size analysis of the generated SPIRV shows that just gathering all generated shaders in the out directory and applying GZip compression yields a 2.58x size reduction with the total size of all generated shaders being about ~48k. This bring the packaged shader size more in line with the other backends. Results of compression of these shaders in an actual APK may vary.

Fixes https://github.com/flutter/flutter/issues/119172
Fixes https://github.com/flutter/flutter/issues/121619
Screenshot 2023-04-02 at 4 08 20 PM

### CODE REVIEW

File: impeller/compiler/compiler.cc [hunk 1]
```diff
@@ -210,309 +207,138 @@ static CompilerBackend CreateCompiler(const spirv_cross::ParsedIR& ir,
...
 
-  SetLimitations(spirv_options);
-
   switch (source_options.target_platform) {
     case TargetPlatform::kMetalDesktop:
-    case TargetPlatform::kMetalIOS:
-      spirv_options.SetOptimizationLevel(
-          shaderc_optimization_level::shaderc_optimization_level_performance);
+    case TargetPlatform::kMetalIOS: {
+      SPIRVCompilerTargetEnv target;
```
Comment (zanderso) [inline impeller/compiler/compiler.cc:254]: It looks like every switch case declares one of these. If you hoist out the declaration you can avoid the `{}` in the cases.

**Review by zanderso — APPROVED** [review]
lgtm w/ optional nit

File: impeller/compiler/spirv_compiler.cc [hunk 2]
```diff
@@ -0,0 +1,305 @@
...
+
+std::string SPIRVCompiler::GetSourcePrefix() const {
+  std::stringstream stream;
+  stream << options_.file_name << ": ";
+  return stream.str();
+}
+
+static void SetDefaultLimitations(shaderc::CompileOptions& compiler_opts) {
+  using Limit = std::pair<shaderc_limit, int>;
+  static constexpr std::array<Limit, 83> limits = {
```
Comment (zanderso) [inline impeller/compiler/spirv_compiler.cc:92]: Not sure why only clang on Windows doesn't like this line. Maybe it wants some explicit `#include`?

File: impeller/compiler/compiler.cc (hunk 1 above)
Comment (chinmaygarde) [reply impeller/compiler/compiler.cc:254]: May I get this in another patch? The compiler needs some TLC. I'm going to get rid of the unknown platforms with optionals and rework this to explicitly reference Stage1 and Stage2.

File: impeller/compiler/spirv_compiler.cc (hunk 2 above)
Comment (chinmaygarde) [reply impeller/compiler/spirv_compiler.cc:92]: Yeah, I moved this to another TU but didn't move the `#include<array>`.

File: impeller/compiler/compiler.cc (hunk 1 above)
Comment (zanderso) [reply impeller/compiler/compiler.cc:253]: Yeah. Still lgtm.

Comment (bdero) [reply impeller/compiler/compiler.cc:253]: Does "Stage 1" mean source->spirv + reflection, and "Stage 2" mean spirv->target (ideally in Vulkan's case, spirv->optimized spirv)?

Comment (chinmaygarde) [reply impeller/compiler/compiler.cc:253]: Yeah, just the diagram in the readme.

**Pull request merged**